### DISCUSSION TITLE
Cannot import FPCA

### DISCUSSION THREAD

**Discussion by Ann-Lilian-Wangui** [GAA-UAM/scikit-fda#529] [scikit-fda discussion]
```python
import skfda
fpca = skfda.preprocessing.dim_reduction.feature_extraction.FPCA(n_components=2)
```
gives
```python
ImportError: cannot import name 'FPCA' from 'skfda.preprocessing.dim_reduction' (/home/ann/.local/lib/python3.9/site-packages/skfda/preprocessing/dim_reduction/__init__.py)
```
and
```python
import skfda
from skfda.datasets import fetch_growth
from skfda.exploratory.visualization import FPCAPlot
from skfda.preprocessing.dim_reduction import FPCA
```
gives
```
ImportError                               Traceback (most recent call last)
<ipython-input-2-1ca6c5a64c6b> in <module>
      2 from skfda.datasets import fetch_growth
      3 from skfda.exploratory.visualization import FPCAPlot
----> 4 from skfda.preprocessing.dim_reduction import FPCA

ImportError: cannot import name 'FPCA' from 'skfda.preprocessing.dim_reduction' (/home/ann/.local/lib/python3.9/site-packages/skfda/preprocessing/dim_reduction/__init__.py)
```
```python
import skfda
skfda.__version__
```
gives
```python
'0.8.1'
```

Running python 3.9.2 on debian 11.

**Comment by vnmabus** [comment]
That is... strange. I cannot reproduce it.
Can you show me the source of the module?:
```python
import inspect
src = inspect.getsource(skfda.preprocessing.dim_reduction)
print(src)
```
It should be:
```python
"""Dim reduction."""
from __future__ import annotations

import importlib
from typing import TYPE_CHECKING, Any

import lazy_loader as lazy

_normal_getattr, __dir__, __all__ = lazy.attach(
    __name__,
    submodules=[
        "variable_selection",
    ],
    submod_attrs={
        "_fpca": ["FPCA"],
    },
)

if TYPE_CHECKING:
    from ._fpca import FPCA as FPCA


def __getattr__(name: str) -> Any:
    if name in {"projection", "feature_extraction"}:
        return importlib.import_module(f".{name}", __name__)
    return _normal_getattr(name)
```

**Comment by Ann-Lilian-Wangui** [comment]
The source is the same as above.

Some upgraded packages in /usr/local/lib/python3.9/dist-packages is breaking this. 
Moving dist-packages to .bak fixes it -- note that sys.path.remove('/usr/local/lib/python3.9/dist-packages') before importing does NOT fix this.

Also, even with the folder first moved, I still had to pip install typing_extensions==4.3.0.

There are still some errors when starting ipython but FPCA can now import:

ann@cairo:~$ cat .ipython/profile_default/startup/skfda_no_usr_local.py 
# To fix skfda.preprocessing.dim_reduction.FPCA
# pip install typing-extensions==4.3.0
import sys
sys.path.remove("/usr/local/lib/python3.9/dist-packages")
sys.path.remove('')

ann@cairo:~$ ipython3
Error processing line 1 of /usr/local/lib/python3.9/dist-packages/matplotlib-3.7.0-py3.9-nspkg.pth:

  Traceback (most recent call last):
    File "/usr/lib/python3.9/site.py", line 175, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.20.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import skfda

In [2]: skfda.preprocessing.dim_reduction.FPCA
Out[2]: skfda.preprocessing.dim_reduction._fpca.FPCA

In [3]:

**Reply by vnmabus** [reply]
That is a bit weird. It sounds like your Python installation has some issues (https://xkcd.com/1987/ 😅). I usually recommend to use a different distribution from your OS Python, such as [miniconda](https://docs.conda.io/en/latest/miniconda.html).

I am glad that you managed to fix your problem at the end!
